### PR TITLE
Bugfix to mbedtls_ssl_parse_new_session_ticket

### PR DESCRIPTION
**Pull request by hannestschofenig**

### CODE REVIEW

File: library/ssl_tls13_generic.c [hunk 1]
```diff
@@ -4643,12 +4643,14 @@ int mbedtls_ssl_parse_new_session_ticket( mbedtls_ssl_context *ssl )
...
     msg_len = ( ssl->in_msg[2] << 8 ) | ssl->in_msg[3];
 
-    if( msg_len+ mbedtls_ssl_hs_hdr_len( ssl ) != ssl->in_msglen )
+    if( ssl->in_hslen == 0 || ssl->in_hslen > ssl->in_msglen )
     {
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad new session ticket message" ) );
         return( MBEDTLS_ERR_SSL_BAD_HS_NEW_SESSION_TICKET );
     }
 
+    ssl->in_msglen -= ssl->in_hslen;
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c:4652]: This seems like a dangerous change, because we change internal fields of the current messaging layer.

@hannestschofenig What's the rationale for this change?

File: library/ssl_tls13_generic.c [hunk 2]
```diff
@@ -4643,12 +4643,14 @@ int mbedtls_ssl_parse_new_session_ticket( mbedtls_ssl_context *ssl )
     msg = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl );
     msg_len = ( ssl->in_msg[2] << 8 ) | ssl->in_msg[3];
 
-    if( msg_len+ mbedtls_ssl_hs_hdr_len( ssl ) != ssl->in_msglen )
+    if( ssl->in_hslen == 0 || ssl->in_hslen > ssl->in_msglen )
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c:4646]: Why are we re-calculating `msg_len` in L.4644? Isn't the length of the HS message stores in `in_hslen` already? Likewise, since we don't support HS fragmentation for now, the messaging layer will have checked that `ssl->in_hslen` isn't greater than `ssl->in_msglen`.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
@hannestschofenig What's the underlying problem that this PR is fixing?

I am confused by both the existing and the new code in `mbedtls_ssl_parse_new_session_ticket()` because it (a) performs HS length checks which are already performed by the underlying messaging layer, as far as I understand, and (b) it modifies the variable `ssl->in_msglen` which is maintained by the messaging layer and shouldn't be modified by the handshake layer.

**Comment by hannestschofenig** [comment]
Closed because it was replaced by #69, which was already merged

**Comment by zhihan** [comment]
Not sure if the issue is fixed. Here is my attempt to connect to enabled.tls13.com:

 % ./ssl_client2 server_name=enabled.tls13.com server_port=443 ca_file=/etc/ssl/cert.pem auth_mode=optional

  . Seeding the random number generator... ok
  . Loading the CA root certificate ... ok (114 skipped)
  . Loading the client cert. and key... ok (key type: EC)
  . Connecting to tcp/enabled.tls13.com/443... ok
  . Setting up the SSL/TLS structure... ok
  . Performing the SSL/TLS handshake... ok
    [ Protocol is TLSv1.3 ]
    [ Ciphersuite is TLS_CHACHA20_POLY1305_SHA256 ]
    [ Key Exchange Mode is ECDHE-ECDSA ]
    [ Record expansion is 5 ]
  . Verifying peer X.509 certificate... failed
  ! The certificate is not correctly signed by the trusted CA

  . Peer certificate information    ...
cert. version     : 3
serial number     : 02:A7:C1:95:57:7F:F0:E9:0C:35:99:61:D2:04:B0:ED
issuer name       : C=US, O=Cloudflare, Inc., CN=Cloudflare Inc ECC CA-3
subject name      : C=US, ST=CA, L=San Francisco, O=Cloudflare, Inc., CN=sni.cloudflaressl.com
issued  on        : 2020-08-17 00:00:00
expires on        : 2021-08-17 12:00:00
signed using      : ECDSA with SHA256
EC key size       : 256 bits
basic constraints : CA=false
subject alt name  :
    dNSName : tls13.com
    dNSName : *.tls13.com
    dNSName : sni.cloudflaressl.com
key usage         : Digital Signature
ext key usage     : TLS Web Server Authentication, TLS Web Client Authentication
certificate policies : ???, ???

  > Write to server: 34 bytes written in 1 fragments

GET / HTTP/1.0
Extra-header:


  < Read from server: mbedtls_ssl_read returned -0x6e00
Last error was: -0x6E00 - SSL - Processing of the NewSessionTicket handshake message failed

**Comment by hanno-becker** [comment]
Thanks @zhihan , there turned out to be another bug not in the way the ticket is parsed, but the way it's fetched from the current messaging layer. Fixed in https://github.com/hannestschofenig/mbedtls/pull/74 and fix confirmed by running the command you pasted.
